lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3630,6 +3628,42 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
                                            ValueWithCleanup(SILValue()));
     if (!insertion.second) // not inserted
       return insertion.first->getSecond();
+
+    // Check whether the original buffer is an address-to-address projection.
+    // If so, recurse until the buffer is such a projection but its operand is
+    // not. Then, get the adjoint buffer of the operand and return a
+    // corresponding projection into it.
+    if (auto *inst = dyn_cast_or_null<SingleValueInstruction>(
+            originalBuffer->getDefiningInstruction())) {
```

Comment:
It'd be simpler to do `Projection::isAddressProjection(originalBuffer)`. This API takes a `SILValue`.

Reply:
Done in https://github.com/apple/swift/pull/22807/commits/d1e0b11e405b7a3bb827788cee68e1e0462f04a8.
Though within the `if` block, I still need to cast to a `SingleValueInstruction`. Less nesting is certainly better though.